syntax = 'proto3';

package api;

import "google/protobuf/timestamp.proto";
import "pkg/api/queue.proto";
import "google/protobuf/empty.proto";
import "github.com/gogo/protobuf/gogoproto/gogo.proto";
import "google/api/annotations.proto";
import "k8s.io/apimachinery/pkg/api/resource/generated.proto";

option (gogoproto.goproto_stringer_all) = false;
option (gogoproto.stringer_all) = true;

// Represents a job being submitted to the queue.
message JobSubmittedEvent {
    string job_id = 1;
    string job_set_id = 2;
    string queue = 3;
    google.protobuf.Timestamp created = 4 [(gogoproto.stdtime) = true, (gogoproto.nullable) = false];
    Job job = 5 [(gogoproto.nullable) = false];
}

// Represents a job being queued.
message JobQueuedEvent {
    string job_id = 1;
    string job_set_id = 2;
    string queue = 3;
    google.protobuf.Timestamp created = 4 [(gogoproto.stdtime) = true, (gogoproto.nullable) = false];
}

// Represents a job is a duplicate of another job.
message JobDuplicateFoundEvent {
    string job_id = 1;
    string job_set_id = 2;
    string queue = 3;
    google.protobuf.Timestamp created = 4 [(gogoproto.stdtime) = true, (gogoproto.nullable) = false];
    string original_job_id = 5;
}

// Represents a job being leased to a K8s cluster.
message JobLeasedEvent {
    string job_id = 1;
    string job_set_id = 2;
    string queue = 3;
    google.protobuf.Timestamp created = 4 [(gogoproto.stdtime) = true, (gogoproto.nullable) = false];
    string cluster_id = 5;
}

// Represents the data for a job that has been leased to a K8s cluster.
message JobLeaseReturnedEvent {
    string job_id = 1;
    string job_set_id = 2;
    string queue = 3;
    google.protobuf.Timestamp created = 4 [(gogoproto.stdtime) = true, (gogoproto.nullable) = false];
    string cluster_id = 5;
    string reason = 6;
    string kubernetes_id = 7;
    int32  pod_number = 8;
}

// Represents k8s saying a job has expired.
message JobLeaseExpiredEvent {
    string job_id = 1;
    string job_set_id = 2;
    string queue = 3;
    google.protobuf.Timestamp created = 4 [(gogoproto.stdtime) = true, (gogoproto.nullable) = false];
}

// Represents a job that is pending in the queue.
message JobPendingEvent {
    string job_id = 1;
    string job_set_id = 2;
    string queue = 3;
    google.protobuf.Timestamp created = 4 [(gogoproto.stdtime) = true, (gogoproto.nullable) = false];
    string cluster_id = 5;
    string kubernetes_id = 6;
    int32 pod_number = 7;
    string pod_name = 8;
    string pod_namespace = 9;
}

// Represents a job that is running
message JobRunningEvent {
    string job_id = 1;
    string job_set_id = 2;
    string queue = 3;
    google.protobuf.Timestamp created = 4 [(gogoproto.stdtime) = true, (gogoproto.nullable) = false];
    string cluster_id = 5;
    string kubernetes_id = 6;
    string node_name = 7;
    int32 pod_number = 8;
    string pod_name = 9;
    string pod_namespace = 10;
}

// Represents the infomation on a job from ingress.
message JobIngressInfoEvent {
    string job_id = 1;
    string job_set_id = 2;
    string queue = 3;
    google.protobuf.Timestamp created = 4 [(gogoproto.stdtime) = true, (gogoproto.nullable) = false];
    string cluster_id = 5;
    string kubernetes_id = 6;
    string node_name = 7;
    int32 pod_number = 8;
    string pod_name = 10;
    string pod_namespace = 11;
    map<int32, string> ingress_addresses = 9;
}

// Represents a job that failed to schedule in the queue.
message JobUnableToScheduleEvent {
    string job_id = 1;
    string job_set_id = 2;
    string queue = 3;
    google.protobuf.Timestamp created = 4 [(gogoproto.stdtime) = true, (gogoproto.nullable) = false];
    string cluster_id = 5;
    string reason = 6;
    string kubernetes_id = 7;
    string node_name = 8;
    int32 pod_number = 9;
    string pod_name = 10;
    string pod_namespace = 11;
}

// Represents a job that failed to run.
message JobFailedEvent {
    string job_id = 1;
    string job_set_id = 2;
    string queue = 3;
    google.protobuf.Timestamp created = 4 [(gogoproto.stdtime) = true, (gogoproto.nullable) = false];
    string cluster_id = 5;
    string reason = 6;
    map<string, int32> exit_codes = 7 [deprecated = true];
    string kubernetes_id = 8;
    string node_name = 9;
    int32 pod_number = 10;
    string pod_name = 13;
    string pod_namespace = 14;
    repeated ContainerStatus container_statuses = 11;
    Cause cause = 12;
}

// Represents a job that has been completed.
message JobSucceededEvent {
    string job_id = 1;
    string job_set_id = 2;
    string queue = 3;
    google.protobuf.Timestamp created = 4 [(gogoproto.stdtime) = true, (gogoproto.nullable) = false];
    string cluster_id = 5;
    string kubernetes_id = 6;
    string node_name = 7;
    int32 pod_number = 8;
    string pod_name = 9;
    string pod_namespace = 10;
}

// Represents the data about a jobs utilization.
message JobUtilisationEvent {
    string job_id = 1;
    string job_set_id = 2;
    string queue = 3;
    google.protobuf.Timestamp created = 4 [(gogoproto.stdtime) = true, (gogoproto.nullable) = false];
    string cluster_id = 5;
    string kubernetes_id = 6;
    map<string, k8s.io.apimachinery.pkg.api.resource.Quantity> MaxResourcesForPeriod = 7 [(gogoproto.nullable) = false];
    string node_name = 8;
    int32 pod_number = 9;
    string pod_name = 10;
    string pod_namespace = 11;
    map<string, k8s.io.apimachinery.pkg.api.resource.Quantity> total_cumulative_usage = 12 [(gogoproto.nullable) = false];
}

// Represents a job that is in the process of reprioritizing.
message JobReprioritizingEvent {
    string job_id = 1;
    string job_set_id = 2;
    string queue = 3;
    google.protobuf.Timestamp created = 4 [(gogoproto.stdtime) = true, (gogoproto.nullable) = false];
    double new_priority = 5;
    string requestor = 6;
}

// Represents a job that has been reprioritized.
message JobReprioritizedEvent {
    string job_id = 1;
    string job_set_id = 2;
    string queue = 3;
    google.protobuf.Timestamp created = 4 [(gogoproto.stdtime) = true, (gogoproto.nullable) = false];
    double new_priority = 5;
    string requestor = 6;
}

// Represents a job that is being cancelled.
message JobCancellingEvent {
    string job_id = 1;
    string job_set_id = 2;
    string queue = 3;
    google.protobuf.Timestamp created = 4 [(gogoproto.stdtime) = true, (gogoproto.nullable) = false];
    string requestor = 5;
}

// Represents a job that has been cancelled.
message JobCancelledEvent {
    string job_id = 1;
    string job_set_id = 2;
    string queue = 3;
    google.protobuf.Timestamp created = 4 [(gogoproto.stdtime) = true, (gogoproto.nullable) = false];
    string requestor = 5;
}

// Represents a job that has been termniated.
message JobTerminatedEvent {
    string job_id = 1;
    string job_set_id = 2;
    string queue = 3;
    google.protobuf.Timestamp created = 4 [(gogoproto.stdtime) = true, (gogoproto.nullable) = false];
    string cluster_id = 5;
    string kubernetes_id = 6;
    int32 pod_number = 7;
    string pod_name = 9;
    string pod_namespace = 10;
    string reason = 8;
}

// Represents a job that is being updated
message JobUpdatedEvent {
    string job_id = 1;
    string job_set_id = 2;
    string queue = 3;
    google.protobuf.Timestamp created = 4 [(gogoproto.stdtime) = true, (gogoproto.nullable) = false];
    string requestor = 5;
    Job job = 6 [(gogoproto.nullable) = false];
}

// Represents at least one of the events from the events.proto file.
message EventMessage {
    oneof events {
        JobSubmittedEvent submitted = 1;
        JobQueuedEvent queued = 2;
        JobDuplicateFoundEvent duplicate_found = 16;
        JobLeasedEvent leased = 3;
        JobLeaseReturnedEvent lease_returned = 4;
        JobLeaseExpiredEvent lease_expired = 5;
        JobPendingEvent pending = 6;
        JobRunningEvent running = 7;
        JobUnableToScheduleEvent unable_to_schedule = 8;
        JobFailedEvent failed = 9;
        JobSucceededEvent succeeded = 10;
        JobReprioritizedEvent reprioritized = 11;
        JobCancellingEvent cancelling = 12;
        JobCancelledEvent cancelled = 13;
        JobTerminatedEvent terminated = 14;
        JobUtilisationEvent utilisation = 15;
        JobIngressInfoEvent ingress_info = 17;
        JobReprioritizingEvent reprioritizing = 18;
        JobUpdatedEvent updated = 19;
    }
}

// The cause for a job to error.
enum Cause {
    Error = 0;
    Evicted = 1;
    OOM = 2;
    DeadlineExceeded = 3;
}

// A containers status.
message ContainerStatus {
    string name = 1;
    int32 exitCode = 2;
    string message = 3;
    string reason = 4;
    Cause cause = 5;
}

// A list of event messages covering all events.
message EventList {
    repeated EventMessage events = 1;
}

// swagger:model
message EventStreamMessage {
    string id = 1;
    EventMessage message = 2;
}

// swagger:model
message JobSetRequest {
    string id = 1;
    bool watch = 2;
    string from_message_id = 3;
    string queue = 4;
    bool errorIfMissing = 5;
    bool force_redis = 6;  // This field is for test purposes only
}

// Represents a request to watch events
message WatchRequest {
    string queue = 1;
    string job_set_id = 2;
    string from_id = 3;
    bool force_redis = 4;  // This field is for test purposes only
}

// Service that handles all events
service Event {

    // Reports multiple events at once
    rpc ReportMultiple (EventList) returns (google.protobuf.Empty);

    // Reports a single event message
    rpc Report (EventMessage) returns (google.protobuf.Empty);

    // Returns an event stream message
    rpc GetJobSetEvents (JobSetRequest) returns (stream EventStreamMessage) {
        option (google.api.http) = {
            post: "/v1/job-set/{queue}/{id}"
            body: "*"
        };
    }

    // Returns an event stream message to watch
    // (deprecated)
    rpc Watch (WatchRequest) returns (stream EventStreamMessage) {
        option deprecated = true;
    }
}
